### DISCUSSION TITLE
Add `groovyls` to `nvim-lspconfig`

### DISCUSSION THREAD

**Discussion by Skylli202** [LazyVim/LazyVim#2089] [Q&A] [answered]
Hi, 

After reading the doc (this [example](https://www.lazyvim.org/configuration/examples) page where there is an example for adding an LSP), trying to copy and adapt one of the LazyExtras for language support I feel like I am very stupid. I love neovim, but what a pain to configure. Many thanks to the contributor of this repo to ease it!

I created a Lua file in my `~/.config/nvim/lua/plugins/` directory, with this content:

```lua
return {
  -- Add groovy to treesitter.
  {
    "nvim-treesitter/nvim-treesitter",
    opts = function(_, opts)
      opts.ensure_installed = opts.ensure_installed or {}
      vim.list_extend(opts.ensure_installed, { "groovy" })
    end,
  },
  {
    "neovim/nvim-lspconfig",
    ---@class PluginLspOpts
    opts = {
      ---@type lspconfig.options
      servers = {
        groovyls = {
          --- that's what the doc of groovy-language-server say to do
          cmd = {
            "java",
            "-jar",
            "~/.local/share/nvim/mason/packages/groovy-language-server/build/libs/groovy-language-server-all.jar",
          },
        },
      },
    },
  },
}

```

When I open a `.groovy` file with neovim, I have no LSP whatsoever.

I have installed the Groovy LSP with Mason. I have Java and Groovy binaries in my path: 

```terminal
➜  nvim git:(main) ✗ java -version
openjdk version "21.0.1" 2023-10-17 LTS
OpenJDK Runtime Environment Corretto-21.0.1.12.1 (build 21.0.1+12-LTS)
OpenJDK 64-Bit Server VM Corretto-21.0.1.12.1 (build 21.0.1+12-LTS, mixed mode, sharing)
➜  nvim git:(main) ✗ groovy -version
Groovy Version: 4.0.15 JVM: 21.0.1 Vendor: Amazon.com Inc. OS: Linux
```

The cmd: `:LspInfo` in the buffer of my `.groovy` file display the following:
```terminal
  Press q or <Esc> to close this window. Press <Tab> to view server doc.
 
 Language client log: /home/skylli/.local/state/nvim/lsp.log
 Detected filetype:   groovy
 
 0 client(s) attached to this buffer: 
 
 Other clients that match the filetype: groovy
 
 Config: groovyls
 	filetypes:         groovy
 	root directory:    Not found.
 	cmd:               /home/skylli/.sdkman/candidates/java/current/bin/java -jar ~/.local/share/nvim/mason/packages/groovy-language-server/build/libs/groovy-language-server-all.jar
 	cmd is executable: true
 	autostart:         true
 	custom handlers:   
 
 Configured servers list: templ, rust_analyzer, groovyls, tailwindcss, eslint, lua_ls, bashls, jsonls, svelte, gopls, tsserver, marksman, taplo
```

For an unknown reason, the LSP does not attach to the buffer, I have probably made a mistake in my groovy lua config. Any help would be greatly appreciated.

**Comment by Skylli202** [comment]
I also tried to set the root directory, as the `LspInfo` state that it is `Not Found`. But no luck that way.

**Comment by dpetka2001** [comment, marked as answer]
I just tried the following
```lua
  {
    "neovim/nvim-lspconfig",
    opts = {
      servers = {
        groovyls = {},
      },
    },
  },
```
As you can see in lspconfig [docs](https://github.com/neovim/nvim-lspconfig/blob/master/lua/lspconfig/server_configurations/groovyls.lua) the server is expecting to find either a `Jenkinsfile` file or a `.git` directory. I don't know anything about groovy and how to create a `Jenkinsfile`, so I just created a `test.groovy` file (given to me by ChatGPT) inside a git repo and the LSP server attached correctly.

**Reply by Skylli202** [reply]
Hi,

Thanks for your time. I missed the part that is saying that the file must be in a git repo. This is the mistake for sure. As my test were conducted in a simple directory with a simple hello world file.

I'll try this at home as soon as I have the opportunity and mark this as resolved hopefully.

**Reply by dpetka2001** [reply]
You could also create a `Jenkinsfile` and it should still attach. Unfortunately I don't know anything at all about that, that's why i tested in a git repo which was easier for me.

**Reply by Skylli202** [reply]
Yeah but I have no knowledge of those files as well. Doing a quick `git init` will be easier!

**Reply by Skylli202** [reply]
Hi @dpetka2001,

I can tell that the original issue is resolved with the configuration you gave and a file in a git repo. The `LspInfo` is properly indicating that one LSP is attached to the buffer.

But I still have an issue, the standard library of Groovy are not unable to be resolved. See the following example:

```groovy
import groovy.json.JsonSlurper 
// Diagnostics:
// unable to resolve class groovy.time.TimeCategory
//  @ line 1, column 1.

String metadata = '{"due-date":"30-Nov-2023","review-expiry-date":"29-Nov-2023"}'
def map = new JsonSlurper().parseText(metadata)
println map.get('review-expiry-date')
```

Executing the previous script with `groovy json.groovy` output the following:
```txt
➜  groovy-sandbox git:(main) ✗ groovy json.groovy
29-Nov-2023
```

The groovy script successfully running indicate that the deps `JsonSlurper` is properly installed on my machine. It is a Linux (fedora, WSL 2) where both Java and Groovy has been installed thought SDKMAN.

**Reply by dpetka2001** [reply]
Sorry, but I have no idea what this is caused by. Maybe a post to the groovy repo will be more helpful.

**Reply by ofgirichardsonb** [reply]
Have you `source`d `$HOME/.sdkman/bin/sdkman-init.sh`?
